content/posts/api-deployment-strategies.md

```diff
@@ -0,0 +1,341 @@
+---
+title: "API Deployment Strategies"
+date: 2023-09-15T09:01:51+05:30
+draft: true
+ShowToc: true
+TocOpen: true
+ShowRelatedContent: true
+summary: "Applying best practices to make new API deployments using the API gateway of choice, Apache APISIX."
+tags: ["apache apisix", "api gateway", "cloud-native"]
+categories: ["API Gateway"]
+cover:
+  image: "/images/api-deployment-strategies/shadow-banner.jpeg"
+  alt: "Shadow in a parking lot."
+  caption: "Deploying changes without affecting users is key to a good deployment strategy."
+  relative: false
+---
+
+Deploying changes to your APIs without impacting clients is vital in most modern software products. In this article, I will focus on some best practices and strategies you can adopt to streamline the deployment of your APIs using API gateways.
+
+Throughout this tutorial, we will use our API gateway of choice, [Apache APISIX](https://apisix.apache.org/), but you can translate these strategies to any other [API gateway](/posts/gateway-and-mesh/) or reverse proxy.
+
+## API Versions
+
+Yes, even that minor change can potentially break your client applications.
+
+API versioning is the most straightforward strategy to make changes to your APIs without disrupting your clients. By versioning, you can let your clients decide whether to upgrade to the new API version. They can also update their applications to the new API at their own pace.
+
+API gateways like APISIX can handle typical scenarios where your clients specify which version of the API to use in their requests. The version is typically specified in one of three ways:
+
+1. In URI path (`/v1/products` and `/v2/products`)
+2. In query parameters (`/products?version=v1` and `/products?version=v2`)
+3. In request headers (`/products -H 'version: v1'` and `/products -H 'version: v2'`)
+
+APISIX can then be configured to route these requests to the appropriate API version.
+
+### In URI Path
+
+The example below shows how you can route traffic to different versions based on the path specified in the URI:
+
+```shell
+curl http://127.0.0.1:9180/apisix/admin/routes/1 -X PUT -d '
+{
+  "uri": "/v1/hello",
```

Comment:
Prefix removal with `proxy-rewrite`? Less you forward `/v1/hello` to the upstream and it doesn't work because it's oblivious to `/v1` (as it should be)

That's the downside of path-based routing, and since you mention plugins afterwards, you should start here IMHO